### ISSUE TITLE
IndexChecker false [array.access.unsafe.high] error

### ISSUE THREAD

**Issue by Maxi17** [typetools/checker-framework#2398]
Consider this snippet:

```
import org.checkerframework.checker.index.qual.IndexOrHigh;
import org.checkerframework.common.value.qual.ArrayLen;
import org.checkerframework.common.value.qual.IntRange;

public class Testing {

    void method(String @ArrayLen(10) [] array, @IndexOrHigh("#1") int index) {
        if (index == 10) {
            index = 0;
        }
        array[index] = "hello"; // Error here
    }

    void methodWithIntRangeParameter(String @ArrayLen(10) [] array, @IntRange(from = 0, to = 10) int index) {
        if (index == 10) {
            index = 0;
        }
        array[index] = "hello";
    }

    void methodWithNegativeTest(String @ArrayLen(10) [] array, @IndexOrHigh("#1") int index) {
        if (index != 10) {
            array[index] = "hello"; // Error here
        }
    }

    void methodWithArrayLengthTest(String @ArrayLen(10) [] array, @IndexOrHigh("#1") int index) {
        if (index != array.length) {
            array[index] = "hello";
        }
    }

    void methodWithUnguardedAssignment(String @ArrayLen(10) [] array, @IndexOrHigh("#1") int index) {
        index = 0;
        array[index] = "hello";
    }
}
```

Running javac -processor org.checkerframework.checker.index.IndexChecker Testing.java should issue 2 errors:

```
Error:(11, 15) java: [array.access.unsafe.high] Potentially unsafe array access: the index could be larger than the array's bound
  found   : @LTEqLengthOf("array") int
  required: @IndexFor("array") or @LTLengthOf("array") -- an integer less than array's length
```

```
Error:(23, 19) java: [array.access.unsafe.high] Potentially unsafe array access: the index could be larger than the array's bound
  found   : @LTEqLengthOf("array") int
  required: @IndexFor("array") or @LTLengthOf("array") -- an integer less than array's length
```

The code is safe because we never get to access 'array' with an invalid index. The if statements before make sure that the index is within bounds.
Comparing the behavior of the first two methods, it looks like the checker can't see that the length of the array is the same as number 10. The next two methods point out something similar: array.length is treated differently from number 10, even though 'array' has been annotated with @ArrayLen(10).
